test/e2e/activator_test.go

```diff
@@ -0,0 +1,175 @@
+func TestActivatorOverload(t *testing.T) {
+	var (
+		logger  = logging.GetContextLogger(t.Name())
+		clients = Setup(t)
+	)
+	const (
+		// The number of concurrent requests to hit the activator with.
+		// 1000 = the number concurrent connections in Istio.
+		concurrency = 1000
+		// Timeout to wait for the responses.
+		// Ideally we must wait ~30 seconds, TODO: need to figure out where the delta comes from.
+		timeout = 65 * time.Second
+		// How long the service will process the request in ms.
+		serviceSleep = 300
+	)
+
+	names := test.ResourceNames{
+		Service: test.AppendRandomString(configName, logger),
+		Image:   "observed-concurrency",
+	}
+
+	configOptions := test.Options{
+		ContainerConcurrency: 1,
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = make(map[string]string)
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations["autoscaling.knative.dev/maxScale"] = "10"
+	}
+
+	test.CleanupOnInterrupt(func() { TearDown(clients, names, logger) }, logger)
+	defer TearDown(clients, names, logger)
+
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(logger, clients, &names, &configOptions, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", domain, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	responseChannel := make(chan *spoof.Response, concurrency)
+
+	sendRequests(client, url, concurrency, responseChannel, timeout, logger, t)
+
+}
+
+func sendRequests(client *spoof.SpoofingClient, url string, concurrency int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) {
+	t.Helper()
+	var (
+		group        errgroup.Group
+		responses    int32
+		wantResponse = http.StatusOK
+	)
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+
+	// Send out the requests asynchronously and wait for them to finish.
+	logger.Info("Starting to send out the requests")
+
+	// Print out stats.
+	go func() {
+		for {
+			select {
+			case <-time.Tick(5 * time.Second):
+				logger.Infof("Received responses: %d", atomic.LoadInt32(&responses))
+			case <-timeoutChan:
+				return
+			}
```

Comment:
Maybe key this off the `errChan` and exit in the ticker once the responses == concurrency?

Reply:
Could you add one more sentence to that?

Reply:
I think you'd want to end this goroutine + the logging before the timeout hits, right?

Reply:
In addition, this will panic, if any of the responses fail, since you'll try to print in the dead `t` object, since you might exit the test before  the `timeoutChan` fires.

Reply:
In case we end up finishing the test before the `timeoutChan` fires  when all responses are received - we return the go routine earlier (see the `if`), in case we encountered errors - either `t.Fatalf` or `t.Errorf` are called. The former would cause the exit of the test, the latter would mark the test as failing and continue the execution. I do not see the case when we would panic here.

Reply:
If there's an error `errChan` below will fire, the test will terminate (Fatalf).
Fatalf will kill the current goroutine, where the test runs, but this is a _different_ goroutine, which will continue to run. 
Test logger, that you construct above, logs to `t.Log[f]`. 
Since it continues to run it will log on ticker and then there will be a panic. Of course it requires timing to be right (e.g. ticker to fire after Fatalf, but before the process terminates killing all goroutines), but this is nonetheless a possibility and we all know about Murphy law. :-).

And I think that is what Markus means above, that you need to exit on errChan first of all, which will make sure that this doesn't happen.

Also given that select executes top to bottom in case all of them match (I'd suggest errChan, timeoutChan, ticker order).

I still stand by my previous comment some time ago that this logging is causing more trouble than it gives us benefit :-)

Reply:
Actually I use the BaseLogger in the go routine, it should not panic. But I will add the exit on the error, it makes sense to prevent the go routine running until the timeout. Thank you.

I think that the case sequence doesn't matter here, the channel is selected randomly when more then one of them signalled.
https://golang.org/ref/spec#Select_statements

Reply:
Hm, I guess it's only `switch` that's top to bottom.
This was an interesting learning experience :)